### ISSUE TITLE
feat: Advanced VRTK_PolicyList

### ISSUE THREAD

**Issue by RunenGamer** [thestonefox/VRTK#1698]
### Environment

    Source of VRTK: Github
    Version of VRTK: release/3.3.0-alpha (e43089a)
    Version of Unity3D (e.g. Unity 2017.3.0f3)
    Hardware used (e.g. Vive)
    SDK used (e.g. SteamVR)

### Steps to reproduce

1. Open "041_Controller_ObjectSnappingToDropZones"
2. See that that "Valid Object List Policy" only supports Tags, Layers and Scripts
3. The scene uses tags to identify correct objects
4. In my game, i got > 100 snap drop zones/objects. Therefore I would need >100 different Tags/Layers/Scripts. 

### Expected behavior

I propose a forth Option for VRTK_PolicyList: "String Identifier"

If you chose VRTK_PolicyList.CheckType = "String Identifier"
Then VRTK_PolicyList.TypeCheck would check if the GameObject obj got an Script VRTK_PolicyListStringIdentifier. If so, it checks whether the string specified in VRTK_PolicyListStringIdentifier matches one identifier of VRTK_PolicyList.

Basically it's the same system like tags, but it don't mess up a projects tag-structure.

If you like it, I can start an PR immeadiatly, because I already programmed this change.

**Comment by moscartong** [comment]
+1. I did something similar in one of my projects. Using a string identity is more convenient when there are lots of different types of objects.

**Comment by bddckr** [comment]
I don't get this feature request I think:

> PolicyList supports tags
> The scene uses tags

???

**Comment by bddckr** [comment]
Additionally:

> PolicyList supports Scripts
> I want VRTK_PolicyList.TypeCheck that checks if the GameObject obj got an Script

This is exactly what the Script functionality does, isn't it?!

**Comment by RunenGamer** [comment]
Not really. The Script functionality checks whether a specified script is attached. 

Example:
You got 100 different Keys with 100 Attachpoints. 
If you do not want to use tags (see above), you can try to use the "script functionality". Therefore you need 100 different scripts (for example "Policy_Key_1.cs" ... "Policy_Key_100.cs").

My solution got only both VRTK_Scripts, where a string (for example "Key_1"..."Key_100") is specified in VRTK_PolicyList_StringIdentifier.


To answer the question:
> Additionally:
>
>  PolicyList supports Scripts
>  I want VRTK_PolicyList.TypeCheck that checks if the GameObject obj got an Script
>
>    This is exactly what the Script functionality does, isn't it?!

When the "new" VRTK_PolicyList wants to check whether the string in VRTK_PolicyList_StringIdentifier is set correctly, it has to check if the script is attached first. It's only a prequirement, not the policy-check itself.

**Comment by bddckr** [comment]
Oh I see. You have more of a 1:1 mapping of items to snap drop zones.

The community is welcome to do a PR, I just would prefer the snap drop zone to be refactored so it just looks for one script and that one then just has a bool returning method on it checks with.

**Comment by RunenGamer** [comment]
This change would not only influence dropZones, because it changes the PolicyList-script, which is widely used within the VRTK-framework. But my suggested changed will be most frequently used for dropZones.

So, is it ok to make a PR request including my suggested changes or should it also refactores the snap drop zones as you descriped?

**Comment by RunenGamer** [comment]
@thestonefox : Since I got a +1 from @m-oscartong and your label "feature" I have your "go for it" to make a PR?

**Comment by thestonefox** [comment]
@RunenGamer sure go for it

**Comment by RunenGamer** [comment]
Ok. I'm working on it. I will make a PR as soon as I'm done.

**Comment by thestonefox** [comment]
We no longer have Policy Lists in V4 instead we have Rules. The `String Identifier` Rule could still be quite useful perhaps.

**Comment by fight4dream** [comment]
is this 
`AnyNameRule` (match name against multiple string) or
`AnyNameMatchRegexRule` (match name against multiple regex) or 
`IsNameMatchRegexRule` (match name against single regex) or
`StringIdentifierRule` (match name against specific regex of the form {prefix="Key_"}{selectIndex=1}) ?
